### ISSUE TITLE
Incorrect handling of tuple literals

### ISSUE THREAD

**Issue by penn5** [microsoft/pyright#4620]
**Describe the bug**
Pyright does not correctly match protocols for addition of tuples if the tuples are literals (maybe there are related issues with different-typed or different-lengthed tuples, I didn't test)

**To Reproduce**
Type check the attached code

**Expected behavior**
The code should pass type checking.
It works if the integer literals are changed to `int("1")` and `int("2")` respectively.

**Screenshots or Code**
```python
import time
import typing
I = typing.TypeVar("I", contravariant=True)
O = typing.TypeVar("O", covariant=True)

class LAddable(typing.Protocol[I, O]):
    def __add__(self, __x: I) -> O:
        ...

class RAddable(typing.Protocol[I, O]):
    def __radd__(self, __x: I) -> O:
        ...

@typing.overload
def slow_add(a: I, b: RAddable[I, O]) -> O:
    ...

@typing.overload
def slow_add(a: LAddable[I, O], b: I) -> O:
    ...

def slow_add(a, b):  # type: ignore
    time.sleep(0.1)
    return a + b

if __name__ == "__main__":
    slow_add((1,), (2,))
```

**VS Code extension or command-line**
pyright 1.1.293

**Additional context**
https://www.reddit.com/r/Python/comments/10zdidm/why_type_hinting_sucks/?utm_medium=android_app&utm_source=share

**Comment by erictraut** [comment]
This is working as designed. I can provide details if you're interested, but I get the sense that you are not a fan of static typing and are probably not interested. In any case, this is not a bug.

I'll note that pyright (unlike mypy or other Python static type checkers) has a special feature called [Call-site Return Type Inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#call-site-return-type-inference). Through this mechanism, pyright can determine the proper type of a call to an untyped function based on the input arguments to the call.

For example:
```python
def add(a, b):
    return a + b

reveal_type(add(1, 2)) # Literal[3]
reveal_type(add(1, 2.0)) # float
reveal_type(add((1, ), (2, ))) # tuple[Literal[1], Literal[2]]]
reveal_type(add(1, 3j)) # complex
reveal_type(add("hi ", "there")) # Literal['hi there']
```

**Comment by penn5** [comment]
> I get the sense that you are not a fan of static typing and are probably not interested. 

Wrong.

In any case, this is not a bug.

> Okay! It's nonetheless very counterintuitive that the only valid signature of `Tuple[Literal[...]].__add__` should be `(Tuple[Literal[...]]) -> Tuple[Literal[...]]`, and I'd be interested to know why this choice was made. I would have put `Tuple[A].__add__: (Tuple[B]) -> Tuple[A, B]` (for single element tuples)
